Python/bytecodes.c

```diff
@@ -2093,61 +2093,42 @@ dummy_func(
             PUSH(len_o);
         }
 
-        // stack effect: (__0, __1 -- )
-        inst(MATCH_CLASS) {
+        inst(MATCH_CLASS, (subject, type, names -- attrs)) {
             // Pop TOS and TOS1. Set TOS to a tuple of attributes on success, or
             // None on failure.
-            PyObject *names = POP();
-            PyObject *type = POP();
-            PyObject *subject = TOP();
             assert(PyTuple_CheckExact(names));
-            PyObject *attrs = match_class(tstate, subject, type, oparg, names);
-            Py_DECREF(names);
-            Py_DECREF(type);
+            attrs = match_class(tstate, subject, type, oparg, names);
+            DECREF_INPUTS();
             if (attrs) {
                 // Success!
                 assert(PyTuple_CheckExact(attrs));
-                SET_TOP(attrs);
             }
             else if (_PyErr_Occurred(tstate)) {
                 // Error!
-                goto error;
+                ERROR_IF(true, error);
             }
             else {
                 // Failure!
-                SET_TOP(Py_NewRef(Py_None));
+                attrs = Py_NewRef(Py_None);
             }
```

Comment:
The `ERROR_IF(true, error)` inside of a conditional feels sort of strange to me. What do you think of this?

```c
            else {
                ERROR_IF(_PyErr_Occurred(tstate), error);
                attrs = Py_NewRef(Py_None);
            }
```

Reply:
TIL that the generator doesn't like trailing comments after an ERROR_IF().

Reply:
Argh, we should fix that. Currently it uses a regex search, which is very lame given that we have an array of tokens available! But moving substitutions to manipulation of the token array is tricky, so improving the regex is probably easier. Example:
```diff
diff --git a/Tools/cases_generator/generate_cases.py b/Tools/cases_generator/generate_cases.py
index d6f9f64ca0..6e433e4cd3 100644
--- a/Tools/cases_generator/generate_cases.py
+++ b/Tools/cases_generator/generate_cases.py
@@ -389,7 +389,7 @@ def write_body(self, out: Formatter, dedent: int, cache_adjust: int = 0) -> None
         assert dedent <= 0
         extra = " " * -dedent
         for line in self.block_text:
-            if m := re.match(r"(\s*)ERROR_IF\((.+), (\w+)\);\s*$", line):
+            if m := re.match(r"(\s*)ERROR_IF\((.+), (\w+)\);\s*(?://.*)?$", line):
                 space, cond, label = m.groups()
                 # ERROR_IF() must pop the inputs from the stack.
                 # The code block is responsible for DECREF()ing them.
```
But there are several other places where this needs to be done (search for `;\s*$`).

Reply:
See 85be319279.

Reply:
Ah sorry I missed your comment.